### PR TITLE
rsmeter: add pubsub datasink 

### LINKED ISSUE #11691: resource-metering: introduce pub/sub mode
## Development Task

Must be done:
- [x] add datasink trait, migrate grpc client to single target  datasink
- [x] centralized place to generate protobuf data
- [x] support multiple datasinks
- [x] recorder separates collector and observer
- [ ] add pubsub datasink #11719

Nice to have:
- [x] a way to shutdown recorder (e.g. RAII).
- [x] `TEST_TAG_PREFIX` is redundant, because testing in rsmeter is isolated. Can remove it.

### PR DESCRIPTION
**Pull request by zhongzc**
### What is changed and how it works?

close #11691 

What's Changed:

* implement pubsub service
* implement pubsub datasink

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
resource_metering: add pubsub mode
```

### CODE REVIEW

File: components/resource_metering/src/reporter/data_sink_reg.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+impl Drop for DataSinkHandle {
+    fn drop(&mut self) {
+        if self.reporter_scheduler.is_none() {
+            return;
+        }
+
+        if let Err(err) = self
+            .reporter_scheduler
+            .as_ref()
+            .unwrap()
```
Comment (andylokandy) [inline components/resource_metering/src/reporter/data_sink_reg.rs:72]: Here is an unwrap

File: components/resource_metering/src/reporter/pubsub.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+        let data_sink = DataSinkImpl { tx };
+        let handle = self.data_sink_reg_handle.register(Box::new(data_sink));
+
+        let report_task = async move {
+            let _h = handle;
+
+            loop {
+                let records = rx.next().await;
+                if records.is_none() {
```
Comment (breeswish) [inline components/resource_metering/src/reporter/pubsub.rs:60]: When will this happen?

File: components/resource_metering/src/reporter/pubsub.rs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                    break;
```
Comment (mornyx) [inline components/resource_metering/src/reporter/pubsub.rs:61]: When will records be `None`? I found the following comment:
```rust
/// Ok(None) when channel is closed and no messages left in the queue
```

But it seems that there is no place to close the channel.

File: components/resource_metering/src/reporter/pubsub.rs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                }
+
+                let _t = REPORT_DURATION_HISTOGRAM.start_timer();
+                let records = records.unwrap();
+                REPORT_DATA_COUNTER
+                    .with_label_values(&["to_send"])
+                    .inc_by(records.len() as _);
+                for record in records.iter() {
```
Comment (breeswish) [inline components/resource_metering/src/reporter/pubsub.rs:69]: When will there be multple records here?

File: components/resource_metering/src/reporter/pubsub.rs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+                    if let Err(err) = sink.send((record.clone(), WriteFlags::default())).await {
+                        warn!("failed to send records"; "error" => ?err);
```
Comment (breeswish) [inline components/resource_metering/src/reporter/pubsub.rs:71]:
```suggestion
                        warn!("failed to send records to the pubsub subscriber"; "error" => ?err);
```

**Review by mornyx — COMMENTED** [review]
The rest LGTM

**Review by breeswish — APPROVED** [review]
the rest LGTM

File: components/resource_metering/src/reporter/pubsub.rs (hunk 3 above)
Comment (zhongzc) [reply components/resource_metering/src/reporter/pubsub.rs:61]: When datasink is dropped, the channel will be closed.

File: components/resource_metering/src/reporter/pubsub.rs (hunk 4 above)
Comment (zhongzc) [reply components/resource_metering/src/reporter/pubsub.rs:69]: Every minute, multiple records will be produced.

File: components/resource_metering/src/reporter/pubsub.rs (hunk 2 above)
Comment (zhongzc) [reply components/resource_metering/src/reporter/pubsub.rs:60]: When datasink is dropped, then tx is dropped, and the channel becomes closed, the routine comes here.

**Review by andylokandy — APPROVED** [review]

**Pull request merged**